streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -246,82 +245,39 @@ public void completeRestoration() {
         }
     }
 
-    /**
-     * <pre>
-     * the following order must be followed:
-     *  1. first close topology to make sure all cached records in the topology are processed
-     *  2. then flush the state, send any left changelog records
-     *  3. then flush the record collector
-     *  4. then commit the record collector -- for EOS this is the synchronization barrier
-     *  5. then checkpoint the state manager -- even if we crash before this step, EOS is still guaranteed
-     * </pre>
-     *
-     * @throws TaskMigratedException if committing offsets failed (non-EOS)
-     *                               or if the task producer got fenced (EOS)
-     */
     @Override
-    public void prepareSuspend() {
-        switch (state()) {
-            case CREATED:
-            case SUSPENDED:
-                // do nothing
-                log.trace("Skip prepare suspending since state is {}", state());
-
-                break;
-
-            case RESTORING:
-                stateMgr.flush();
-                log.info("Prepare suspending restoring");
-
-                break;
-
-            case RUNNING:
-                closeTopology(true);
-
-                stateMgr.flush();
-                recordCollector.flush();
-
-                log.info("Prepare suspending running");
-
-                break;
-
-            case CLOSED:
-                throw new IllegalStateException("Illegal state " + state() + " while suspending active task " + id);
-
-            default:
-                throw new IllegalStateException("Unknown state " + state() + " while suspending active task " + id);
-        }
+    public void suspendDirty() {
+        log.info("Suspending dirty");
+        suspend(false);
     }
 
     @Override
-    public void suspend() {
+    public void suspendCleanAndPrepareCommit() {
+        log.info("Suspending clean");
+        suspend(true);
+    }
+
+    @SuppressWarnings("fallthrough")
+    private void suspend(final boolean clean) {
```

Comment:
The "old" suspend() was called _after_ committing, the "new" suspend() is now called _before_ committing!

The old suspend logic is now handled via `postCommit` and `close`